rust/connlib/tunnel/src/tests.rs

```diff
@@ -590,27 +599,68 @@ impl ReferenceStateMachine for ReferenceState {
+                                    prev_ips.retain(|(prev_ip, _)| {
+                                        if prev_ip == ip {
+                                            tracing::debug!(
+                                                "Expiring DNS record: '{prev_domain} -> {prev_ip}'"
+                                            );
+                                            return false;
+                                        }
+
+                                        true
+                                    });
                                 }
-                                _ => unimplemented!(),
                             }
-                        });
 
-                        state
-                            .client_dns_records
-                            .entry(domain.clone())
-                            .or_default()
-                            .extend(ips_resolved_by_query);
+                            // We only add new records when we first connect to a resource.
+                            for ip in all_ips {
+                                state
+                                .client_dns_records
+                                .entry(domain.clone())
+                                .or_default()
+                                .push({
+                                    let ip_status = {
+                                        #[allow(clippy::wildcard_enum_match_arm)]
+                                        match r_type {
+                                            RecordType::A if ip.is_ipv4() => IpStatus::Queried,
+                                            RecordType::AAAA if ip.is_ipv6() => IpStatus::Queried,
+                                            RecordType::A | RecordType::AAAA => IpStatus::Internal,
+                                            _ => unimplemented!(),
+                                        }
+                                    };
+
+                                    tracing::debug!(
+                                        "Adding DNS record on client: '{domain} -> {ip}' ({ip_status:?})"
+                                    );
+
+                                    (ip, ip_status)
+                                });
+                            }
+
+                            state.client_dns_records.entry(domain).or_default().sort();
+                        } else {
+                            // If we are already connected, any existing entries are switched to `Queried` to make the test harness aware that we can now send packets to this address.
+                            for (ip, status) in state.client_dns_records.entry(domain).or_default()
+                            {
+                                let new_status = {
+                                    #[allow(clippy::wildcard_enum_match_arm)]
+                                    match r_type {
+                                        RecordType::A if ip.is_ipv4() => IpStatus::Queried,
```

Comment:
So this is the case you mentioned where the user previously made an AAAA query and we also got an IPv4 address and marked it Internal, then they made an A query and we upgrade it to Queried.